File: 2213-LongestSubstringofOneRepeatingCharacter/2213-LongestSubstringofOneRepeatingCharacter.py
# Last updated: 8/13/2026, 1:30:10 PM
1from typing import *
2from collections import defaultdict, deque, Counter
3from functools import lru_cache, cache
4from itertools import accumulate, combinations, permutations, product
5from bisect import bisect_left, bisect_right, insort
6from heapq import *
7from math import *
8from string import ascii_lowercase, ascii_uppercase
9from sortedcontainers import SortedList
10import random
11import re
12
13class Node:
14    def __init__(self, left=0, right=0, best=0):
15        self.left = left
16        self.right = right
17        self.best = best
18
19class SegmentTree:
20    def __init__(self, nums):
21        self.nums = nums
22        self.n = len(nums)
23        self.m = self.n * 4
24
25        self.indexes = [-1 for _ in range(self.n)]
26        self.tree = [None for _ in range(self.m)]
27        self.build(0, 0, self.n-1)
28
29    def mergeNodes(self, left, right, l, mid, r):
30        node = Node(left.left, right.right, max(left.best, right.best))
31
32        if self.nums[mid] == self.nums[mid + 1]:
33            node.best = max(node.best, left.right + right.left)
34
35            if left.left == (mid - l + 1):
36                node.left = left.left + right.left
37            if right.left == (r - mid):
38                node.right = right.left + left.right
39
40        return node
41
42    def build(self, index, l, r):
43        if l == r:
44            self.tree[index] = Node(1, 1, 1)
45            self.indexes[l] = index
46            return self.tree[index]
47
48        mid = (l + r) // 2
49
50        left = self.build(index * 2 + 1, l, mid)
51        right = self.build(index * 2 + 2, mid + 1, r)
52
53        self.tree[index] = self.mergeNodes(left, right, l, mid, r)
54        return self.tree[index]
55
56    def update(self, index, l, r, left, right):
57        if r < left or l > right:
58            return self.tree[index]
59
60        if l >= left and r <= right:
61            return self.tree[index]
62
63        mid = (l + r) // 2
64        left_node = self.update(index * 2 + 1, l, mid, left, right)
65        right_node = self.update(index * 2 + 2, mid + 1, r, left, right)
66
67        self.tree[index] = self.mergeNodes(left_node, right_node, l, mid, r)
68        return self.tree[index]
69
70class Solution:
71    def longestRepeating(self, s: str, queryCharacters: str, queryIndices: List[int]) -> List[int]:
72        n = len(s)
73        k = len(queryCharacters)
74        s = list(s)
75
76        ans = []
77        segTree = SegmentTree(s)
78
79        for i in range(k):
80            index = queryIndices[i]
81            s[index] = queryCharacters[i]
82
83            node = segTree.update(0, 0, n-1, index, index)
84            ans.append(node.best)
85
86        return ans
87

File: 2213-LongestSubstringofOneRepeatingCharacter/2213-LongestSubstringofOneRepeatingCharacter.cpp
// Last updated: 4/9/2025, 4:02:10 pm
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int maxLen = 1;
    int left = 1;
    int right = 1;
    char leftChar = 0;
    char rightChar = 0;
    int window = 1;
};

class SegmentTree {
public:
    int n;
    string alp;
    vector<Node> tree;
    vector<int> indexes;

    SegmentTree(const string &s) {
        alp = s;
        n = (int)s.size();
        tree.resize(4 * n);
        indexes.resize(n);
        build(0, n - 1, 0);
    }

    void assignTreeIndex(int index, const Node &left, const Node &right) {
        Node &cur = tree[index];
        cur.leftChar = left.leftChar;
        cur.rightChar = right.rightChar;

        if (left.rightChar == right.leftChar) {
            int maxCnt = left.right + right.left;
            cur.maxLen = max({left.maxLen, right.maxLen, maxCnt});

            if (left.left == left.window)
                cur.left = maxCnt;
            else
                cur.left = left.left;

            if (right.right == right.window)
                cur.right = maxCnt;
            else
                cur.right = right.right;
        } else {
            cur.maxLen = max(left.maxLen, right.maxLen);
            cur.left = left.left;
            cur.right = right.right;
        }
    }

    Node build(int l, int r, int index) {
        if (l == r) {
            indexes[l] = index;
            tree[index].leftChar = alp[l];
            tree[index].rightChar = alp[l];
            tree[index].window = 1;
            tree[index].maxLen = 1;
            tree[index].left = 1;
            tree[index].right = 1;
            return tree[index];
        }

        int mid = (l + r) / 2;
        Node left = build(l, mid, index * 2 + 1);
        Node right = build(mid + 1, r, index * 2 + 2);

        tree[index].window = left.window + right.window;
        assignTreeIndex(index, left, right);

        return tree[index];
    }

    void update(int pos, char newVal) {
        int index = indexes[pos];
        tree[index].leftChar = newVal;
        tree[index].rightChar = newVal;
        tree[index].maxLen = 1;
        tree[index].left = 1;
        tree[index].right = 1;

        while (index > 0) {
            index = (index - 1) / 2;
            Node &left = tree[index * 2 + 1];
            Node &right = tree[index * 2 + 2];
            tree[index].window = left.window + right.window;
            assignTreeIndex(index, left, right);
        }
    }
};

class Solution {
public:
    vector<int> longestRepeating(string s, string queryCharacters, vector<int> &queryIndices) {
        int n = queryCharacters.size();
        SegmentTree segTree(s);
        vector<int> ret;
        ret.reserve(n);

        for (int i = 0; i < n; i++) {
            segTree.update(queryIndices[i], queryCharacters[i]);
            ret.push_back(segTree.tree[0].maxLen);
        }
        return ret;
    }
};
